File: cpp/0744.0_Find_Smallest_Letter_Greater_Than_Target.cpp
/*
binary search, T: O(logN), S: O(1)

执行用时：24 ms, 在所有 C++ 提交中击败了11.35% 的用户
内存消耗：15.4 MB, 在所有 C++ 提交中击败了73.23% 的用户
通过测试用例：165 / 165
*/
class Solution {
public:
    char nextGreatestLetter(vector<char>& letters, char target) {
        int len = letters.size();
        int l = 0, r = len - 1;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (target < letters[mid])
                r = mid - 1;
            else
                l = mid + 1;
        }
        return letters[l % len];
    }
};


/*
执行用时：8 ms, 在所有 C++ 提交中击败了97.62% 的用户
内存消耗：15.4 MB, 在所有 C++ 提交中击败了64.13% 的用户
通过测试用例：165 / 165
*/
class Solution {
public:
    char nextGreatestLetter(vector<char>& letters, char target) {
        int idx = upper_bound(letters.begin(), letters.end(), target) - letters.begin();
        return letters[idx % letters.size()];
    }
};


/*
执行用时：12 ms, 在所有 C++ 提交中击败了79.96% 的用户
内存消耗：15.3 MB, 在所有 C++ 提交中击败了92.05% 的用户
通过测试用例：165 / 165
*/
class Solution {
public:
    char nextGreatestLetter(vector<char>& letters, char target) {
        int n = letters.size();
        int l = 0, r = n;
        while (l < r) {
            int mid = l + (r - l) / 2;
            if (letters[mid] <= target)
                l = mid + 1;
            else
                r = mid;
        }
        return letters[l % n];
    }
};

File: leetcode-cn/0744.0_Find_Smallest_Letter_Greater_Than_Target.py
'''
binary search, T: O(logN), S: O(1)

执行用时：28 ms, 在所有 Python3 提交中击败了97.20% 的用户
内存消耗：16.9 MB, 在所有 Python3 提交中击败了51.85% 的用户
通过测试用例：165 / 165
'''
class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        n = len(letters)
        idx = bisect.bisect_right(letters, target)
        if idx == n:
            return letters[0]
        else:
            return letters[idx]

'''
binary search, T: O(logN), S: O(1)

执行用时：44 ms, 在所有 Python3 提交中击败了30.97% 的用户
内存消耗：17 MB, 在所有 Python3 提交中击败了34.67% 的用户
通过测试用例：165 / 165
'''
class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        return letters[bisect.bisect_right(letters, target) % len(letters)]

